proposals/0000-quick-look-impredicativity.md

```diff
@@ -0,0 +1,146 @@
+
+One important feature of Haskell's type system that "quick look" uses is the invariance of type constructors. In short, the subsumption rules ensure that if, for example, `Maybe t` is a more polymorphic than `Maybe s`, it must be the case that `t` equals `s`. This property holds for every type constructor except for function types, which require a slightly more complex handling. The proposed inference algorithm never tries to perform any complicated analysis on other types: impredicativity must be the *only obvious* solution to make the program type check (or as we say in the paper, "impredicativity is never guessed"). Note that since [proposal 287, *Simplify Subsumption*](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0287-simplify-subsumption.rst) has been accepted, function types are also considered invariant.
+
+## Examples
+
+Several examples can be found in the [paper draft](https://www.microsoft.com/en-us/research/publication/a-quick-look-at-impredicativity/). Let us review the main example, `(\x -> x) : ids`, where `ids :: [forall a. a -> a]` from the eyes of the "approachable" description.
+
+1. We infer the type of the head of the application, `(:) :: forall a. a -> [a] -> [a]`.
+2. We expose two arguments, which leads to the type `alpha -> [alpha] -> [alpha]` where `alpha` is a fresh unification variable.
+3. Now we do *quick look* by checking the simple expressions against those types:
+  a. `\x -> x` is not simple, so nothing is done.
+  b. `ids :: [forall a. a -> a]` is checked against `[alpha]`. Since `alpha` is under a type constructor different than arrow, `alpha` *must* be `forall a. a -> a` for the expression to type check.
+4. The result of quick look is thus `alpha := forall a. a -> a`. This means that the second, real type checking phase must check:
+  a. `\x -> x` against `forall a. a -> a`,
+  b. `ids` against `[forall a. a -> a]`.
+
+Compare this example to `(\x -> x) : []`. In this case quick look does not return any impredicativity information (since `[]` does not contain any). Thus the inferred type for that expression is `[beta -> beta]`, without any impredicative polymorphism.
+
+## Effect and Interactions
+
+As discussed several times throughout this proposal, its goal is to give a clear and simple specification of impredicative instantiation within GHC. Whereas "simple" is a subjective matter, the availability of a set of typing rules defines a clear specification.
+
+### `TypeApplications`
+
+We deem the interaction with `TypeApplications` as a very important one; the goal is for our specification to benefit from user-written types as much as possible. We want `(:) @(forall a. a -> a) (\x -> x) []` to work, without the need of an additional type application in `[]`. The draft paper details the interaction between those features in depth.
+
+### The typing rule for `($)`
```

Comment:
Are you or are you not proposing that the `($)` rule be dropped? This section (and the corresponding alternative) don't unclear to me.

Reply:
I have reworded this a bit. I propose to remove the rule. However, I want to make clear that due to GHC's lenient behavior, things like `runST $ ...` which only required `RankNTypes` now may require `ImpredicativeTypes`.